### ISSUE TITLE
How to add negative samples for training

### ISSUE THREAD

**Issue by meng8407** [RangiLyu/nanodet#300]
Thanks for sharing，
I want to add some negative samples as background training, how should I do it? Could you give me a detailed introduction? 
I try to add some empty tag data to the training set and then train, is that ok？
Looking forward to your reply，Thank you

**Comment by raisinglc** [comment]
I do not know how to add neg when using default coco api, so I assume you are using the xml one. You can just add some data with a category of 'NegSample', which does not present in the class name list. When the the dataloader load a sample, whose category does not present in the class list, the dataloader will give a result with zero as its label ID(neg sample).

**Comment by meng8407** [comment]
> I do not know how to add neg when using default coco api, so I assume you are using the xml one. You can just add some data with a category of 'NegSample', which does not present in the class name list. When the the dataloader load a sample, whose category does not present in the class list, the dataloader will give a result with zero as its label ID(neg sample).

Thank you very much for your reply. 
I used the coco format data made by myself. According to your suggestion, when I make coco format data, I can  add some data with a category of 'NegSample', which does not present in the class name list. will the data loader give a result with tag ID 0?Will it be trained as a negative sample?
Looking forward to your reply，Thank you!

**Comment by raisinglc** [comment]
Yes, I think it works. I am using my own json dataloader, but I made it following the xml dataloader.

In the following two pieces of code, you pass a sample with category of 'negsample'(the name could be anyone you want), the data processor returns an array of zero which means there is no target object in this frame, in other word this sample is an negative sample. 

```
#  xml_dataset

for idx, xml_name in enumerate(ann_file_names):
            tree = ET.parse(os.path.join(ann_path, xml_name))
            root = tree.getroot()
            file_name = root.find('filename').text
            width = int(root.find('size').find('width').text)
            height = int(root.find('size').find('height').text)
            info = {'file_name': file_name,
                    'height': height,
                    'width': width,
                    'id': idx + 1}
            image_info.append(info)
            for _object in root.findall('object'):
                category = _object.find('name').text
                if category not in self.class_names:
                    logging.warning("WARNING! {} is not in class_names! Pass this box annotation.".format(category))
                    continue
                for cat in categories:
                    if category == cat['name']:
                        cat_id = cat['id']
                xmin = int(_object.find('bndbox').find('xmin').text)
                ymin = int(_object.find('bndbox').find('ymin').text)
                xmax = int(_object.find('bndbox').find('xmax').text)
                ymax = int(_object.find('bndbox').find('ymax').text)
                w = xmax - xmin
                h = ymax - ymin
                if w < 0 or h < 0:
                    logging.warning("WARNING! Find error data in file {}! Box w and h should > 0. Pass this box "
                                    "annotation.".format(xml_name))
                    continue
                coco_box = [max(xmin, 0), max(ymin, 0), min(w, width), min(h, height)]
                ann = {'image_id': idx + 1,
                       'bbox': coco_box,
                       'category_id': cat_id,
                       'iscrowd': 0,
                       'id': ann_id,
                       'area': coco_box[2] * coco_box[3]
                       }
                annotations.append(ann)
                ann_id += 1

```

```
# coco.py

if gt_bboxes:
            gt_bboxes = np.array(gt_bboxes, dtype=np.float32)
            gt_labels = np.array(gt_labels, dtype=np.int64)
        else:
            gt_bboxes = np.zeros((0, 4), dtype=np.float32)
            gt_labels = np.array([], dtype=np.int64)
```

**Comment by meng8407** [comment]
Thank you again for your reply. 
One more question: can the empty label mentioned on the Internet be regarded as a negative sample? That is, if there is the position of the target object, but the label is empty, Can we also think of it as a negative sample of training?

**Comment by raisinglc** [comment]
I think the answer is yes, if the label is "" in the annotation file. I don't know whether a file without a category is valid in the coco processor.  The data processor would treat is as negative sample, because "" doesn't present in the class list.

**Comment by meng8407** [comment]
> I think the answer is yes, if the label is "" in the annotation file. I don't know whether a file without a category is valid in the coco processor. The data processor would treat is as negative sample, because "" doesn't present in the class list.


Thank you for your reply and wish you all the best!
